cassiopeia/solver/CCPhyloSolver.py

```diff
@@ -0,0 +1,223 @@
+"""
+This file stores a subclass of CassiopeiaSolver, the CCPhyloSolver. This is 
+a wrapper around optimized agglomerative algorithms implemented by CCPhylo
+(https://bitbucket.org/genomicepidemiology/ccphylo/src/master/). Methods
+that will inherit from this class by default are DynamicNeighborJoiningSolver,
+HeuristicNeighborJoiningSolver, NeighborJoiningSolver and UPGMASolver.
+"""
+
+import abc
+from typing import Callable, Dict, List, Optional, Tuple
+
+import networkx as nx
+import numpy as np
+import os
+import tempfile
+import ete3
+import subprocess
+import pandas as pd
+import configparser
+
+from cassiopeia.data import CassiopeiaTree, utilities
+from cassiopeia.mixins import DistanceSolverError
+from cassiopeia.solver import (
+    DistanceSolver,
+    dissimilarity_functions,
+    solver_utilities,
+)
+
+
+class CCPhyloSolver(DistanceSolver.DistanceSolver):
+    """
+    Distance based solver class. This solver serves as a wrapper around 
+    CCPhylo algorithms.
+
+    Args:
+        dissimilarity_function: Function that can be used to compute the
+            dissimilarity between samples.
+        add_root: Whether or not to add an implicit root the tree. Only
+            pertinent in algorithms that return an unrooted tree, by default
+            (e.g. Neighbor Joining). Will not override an explicitly defined
+            root, specified by the 'root_sample_name' attribute in the
+            CassiopeiaTree
+        prior_transformation: Function to use when transforming priors into
+            weights. Supports the following transformations:
+                "negative_log": Transforms each probability by the negative
+                    log (default)
+                "inverse": Transforms each probability p by taking 1/p
+                "square_root_inverse": Transforms each probability by the
+                    the square root of 1/p
+
+    Attributes:
+        dissimilarity_function: Function used to compute dissimilarity between
+            samples.
+        add_root: Whether or not to add an implicit root the tree.
+        prior_transformation: Function to use when transforming priors into
+            weights.
+    """
+
+    def __init__(
+        self,
+        dissimilarity_function: Optional[
+            Callable[
+                [np.array, np.array, int, Dict[int, Dict[int, float]]], float
+            ]
+        ] = dissimilarity_functions.weighted_hamming_distance,
+        add_root: bool = False,
+        prior_transformation: str = "negative_log",
+        fast = True, 
+        method: str = "nj",
+    ):
+
+        super().__init__(prior_transformation)
+
+        self.dissimilarity_function = dissimilarity_function
+        self.add_root = add_root
+        self.fast = fast
```

Comment:
I suggest we call `_setup_ccphylo` in the constructor if `fast=True`.  The reason being is that I want to alert the user as soon as possible if there's an error in the configuration, and that I don't imagine the configuration changing between successive runs of the solver.

Reply:
Good point! I've fixed this.